### DISCUSSION TITLE
Help Needed: Data Cleaning Challenge

### DISCUSSION THREAD

**Discussion by a-y-u-s-h** [saulpw/visidata#1147] [Q&A] [answered]
Hi,

I recently started learning Visidata and I figured best way to practice would be to clean some data. So while looking for dirty data, I stumbled upon [this data munging challenge video](https://www.youtube.com/watch?v=sSnbmbRmtSA) and I'm having a hard time finding solution to this with Visidata. I spent a few hours digging up the documentation, but I couldn't find the section that tells me what variables we can use within python expressions. I have a hunch that if I can get access to any cell from any other cell in the sheet, I should be able to find my way with cleaning this data that's shown in the video. Requesting help from Visidata ninjas. Would appreciate any help or hint with this. Complete solutions are fine as well, hints are also appreciated.

**Comment by saulpw** [comment, marked as answer]
Hey @a-y-u-s-h, thanks for sharing!  I love challenges like this.  I didn't watch the whole video, but it seems like it's pretty straightforward.  Load up the csv in VisiData, move the cursor down to an "Iter" row and press `,` to select all Iter rows (or you can `|Iter` to select without moving the cursor), and then press `"` to pull all selected rows into a separate sheet.  Press `-` on the blank columns at the end (and any other columns you want to remove).  Then press `#` on each column to type it as `int`; or, something like `C gs g#` to do it in bulk from the Columns Sheet.

I think that's all he's doing in the video, but with more steps and code :)  Is there anything else that I missed?

**Reply by a-y-u-s-h** [reply]
Yeah, actually it's a bit different. So the input data looks like this: 

![screenshot_20211009_043302](https://user-images.githubusercontent.com/19586719/136633106-ea48f730-7559-4e9e-8bfb-28e2db294c14.png)

The problem actually shows dummy data. Each row-separated table belongs to a separate individual. In the final output he has 3 more columns for first name, last name and date (in the input that information is written between the blank-row-separated tables). But also in input, he has `first name: Person`, `last name: Human` and `date: end of time` which is common among all rows, but he says that we need to ignore that they're common or that we should process the file assuming that one table belongs to one person (who is not identical to another one), so the output should look like: 

![screenshot_20211009_043324](https://user-images.githubusercontent.com/19586719/136633481-7b054937-3208-461b-99cb-5a96ee9b459d.png)

First 11 rows are going to be about first person. The first 3 columns will contain identity of the first person (info written on top of their table) and it will be same for first 11 rows. Then next 11 rows will contain table of second person and so on. So I figured, we'd need 3 more columns (`za`) and then I need to input some sort of python expression that can access any of the sheet's cells so that I can pick the cells that I want (by using some modulo operator to find out the index of first name for the person, and the number of rows that they need to fill).

Also, `Average`, `Maximum` and `Std.Dev` are some more properties specific to the individual which is not calculated from the `Iter` columns, so we need to get that information as well. My partial solution was to make 3 columns in existing sheet format, select and delete rows that are not needed and then try to use some python-expression trickery on those 3 columns to clean it, but I don't have access to specific cells in them I think.

**Reply by saulpw** [reply]
Ah, I see.  Dummy data always make things less clear.  The key command you want here is `f` to fill null values with the rows above.  Basically:
 - copy those 3 columns with `'` ("freeze")
 - select all title rows with `|first name` and then invert the selection with `gt`
 - go to one of the frozen columns and `gzd` to null the values in the selected rows (everything but the title rows)
 - `f` to fill the null values with the closest value above
 - repeat for each frozen column
 - then you can delete/rename the other unwanted rows/columns as mentioned previously.

I don't quite understand what you want to do with the Average/Max/StdDev rows though.  Can you clarify that?

**Reply by a-y-u-s-h** [reply]
Thanks, damn this was clever. Made it so easy. I won't forget this copying column trick now. This is how the final output looks like now after a few more regex based clean ups (how it should be) - 

 
![screenshot_20211009_050938](https://user-images.githubusercontent.com/19586719/136635033-1251e082-348a-4369-8b8d-13046fab2405.png)

What I meant by that was, that `Average/Max/StdDev` rows also needed to be present in the final output under `Row Type` column and selecting `Iter` only would not include them. They're also not calculated from the fields above them, so they also needed to be copied. But your method took care of that. I checked this by modifying some first few person names as well, works as wanted.

**Reply by saulpw** [reply]
Awesome, glad we found an approach that works!
